corehq/apps/translations/app_translations/upload_app.py

```diff
@@ -248,42 +248,58 @@ class BulkAppTranslationModulesAndFormsUpdater(BulkAppTranslationUpdater):
+            raise ValueError
+        module_index = int(identifying_parts[0].replace("menu", "").replace("module", "")) - 1
+        document = self.app.get_module(module_index)
+        if len(identifying_parts) == 2:
+            form_index = int(identifying_parts[1].replace("form", "")) - 1
+            document = document.get_form(form_index)
+        return document
+
+    def get_document_by_identifier(self, identifier):
+        unique_id = identifier.split(':')[1]
+        if is_module_sheet(identifier):
+            return self.app.get_module_by_unique_id(unique_id)
+        if is_form_sheet(identifier):
+            return self.app.get_form(unique_id)
+
+    def get_document(self, identifier):
+        message = None
+        try:
+            if is_legacy_module_sheet(identifier) or is_legacy_form_sheet(identifier):
+                document = self.get_document_by_legacy_identifier(identifier)
+            else:
+                document = self.get_document_by_identifier(identifier)
+        except (IndexError, ValueError) as err:
+            message = _('Did not recognize "%s", skipping row.') % identifier
+        except ModuleNotFoundException as err:
+            message = _('Invalid menu in row "%s", skipping row.') % identifier
+        except FormNotFoundException as err:
+            message = _('Invalid form in row "%s", skipping row.') % identifier
+        if message:
+            if six.PY2:
+                err.message = message
+                raise err
+            else:
+                raise err.__class__(message) from err
+        return document
+
     def update(self, rows):
         """
         This handles updating module/form names and menu media
         (the contents of the "Menus and forms" sheet in the multi-tab upload).
         """
         self.msgs = []
         for row in get_unicode_dicts(rows):
-            identifying_text = row.get('menu_or_form', row.get('sheet_name', ''))
-            identifying_parts = identifying_text.split('_')
-
-            if len(identifying_parts) not in (1, 2):
-                self.msgs.append((
-                    messages.error,
-                    _('Did not recognize "%s", skipping row.') % identifying_text
-                ))
-                continue
-
-            module_index = int(identifying_parts[0].replace("menu", "").replace("module", "")) - 1
+            identifier = row.get('menu_or_form', row.get('sheet_name', ''))
             try:
-                document = self.app.get_module(module_index)
-            except ModuleNotFoundException:
-                self.msgs.append((
-                    messages.error,
-                    _('Invalid menu in row "%s", skipping row.') % identifying_text
-                ))
+                document = self.get_document(identifier)
+            except (IndexError, ValueError, ModuleNotFoundException, FormNotFoundException) as err:
+                self.msgs.append((messages.error, str(err)))
```

Comment:
Shouldn't `str(err)` be `six.text_type(err)`? I don't have a rationale for this, I just think I've seen the latter in our code more often than the former.